docs/libvirt-node-networking.md

```diff
@@ -0,0 +1,298 @@
+    is.
+ 2) How can the nested nodes and the real node communicate with each other.
+ 3) How to bring a device in position, so that libvirt can use it.
+ 
+ Further the proposal focuses on two network topologies:
+ 
+ 1) The master host network device is **not** a **bridge**
+ 2) The master host network device is a **bridge**.
+ 
+ In both cases, the goal is to keep the host network as much as it is. Thus the
+ most optimal solution only adds network related object (device, routes, ...),
+ but does not modify existing configurations.
+
+ For both network topologies, the existence of a DHCP service for the nodes
+ network is required.
+
+## How to specify the host network device
+
+Right now, adding an extra network type `NodeNetwork` will be added. It does
+not take any extra arguments. On the node, the CNI plugin which creates a new
+endpoint for the device will get the node device name passed in by
+virt-handler. Virt-handler, which has to invoke the plugin from the host
+network namespace view, will fetch the node IP from the `Node` object in the
+cluster, and derive the device from it.
+
+The section on the `Node` from which the IP needs to be taken:
+
+```yaml
+status:
+  addresses:
+  - address: 192.168.121.180
+    type: InternalIP
+```
+
+Based on that IP we can derive the ethernet device which we need. A simple bash
+script like
+```bash
+DEV=$(ip route | grep "src 192.168.121.180" | sed -e "s/.*dev \([^ ]*\).*/\1/g")
+echo $DEV
+eth0
+```
+illustrates that.
+
+Additionally annotations or labels can be considered, to override the detection
+procedure, and directly use a specified device.
+
+## Non-Bridge Scenario (macvlan/macvtap)
+
+```
+  Host               Container                               VM
+| ---------------- | ------------------------------------- | ------------ |
+|           KubeVirt CNI          |                     Libvirt           |
+| ------------------------------- | --------------------------------------| 
+
+  eth0 ----------------macvlan1 -------> macvtap1 -------------> /dev/ethX
+```
+
+virt-handler calls a KubeVirt CNI plugin. The plugin creates a macvlan
+interface, moves it into the container, and attaches the macvlan interface to
+the host interface. The plugin is also responsible for generating/providing MAC
+addresses, if none are provided. The plugin returns the name of the veth
+endpoint in the container, and the MAC address. Libvirt then uses it's `direct`
+attach capability, by creating a macvtap interface, and configures it with the
+MAC address provided.
+
+To check: 
+
+ * Does libvirt need udev in the container or does it the work itself?
+
+Advantage:
+
+ * At the end macvtaps are used, which means that the VM get's the mac address
+   assigned by libvirt
+ * In theory macvlans and macvtaps are faster than bridges.
+ * We can do networking this way, on pretty much every k8s cluster out there,
+   without any specific needs on the host network.
```

Comment:
Only if the assumption holds that the `internalIP` is really naming the device used for connecting containers.